libcxx/test/std/numerics/numeric.ops/numeric.iota/ranges.iota.pass.cpp

```diff
@@ -0,0 +1,166 @@
+
+//
+// Testing results
+//
+
+struct DangerousCopyAssign {
+  int val;
+  using difference_type = int;
+
+  constexpr explicit DangerousCopyAssign(int v) : val(v) {}
+
+  // Needed in postfix
+  constexpr DangerousCopyAssign(DangerousCopyAssign const& other) { this->val = other.val; }
+
+  // mischievious copy assignment that we won't use if the
+  // std::as_const inside ranges::iota isn't working, this should perturb the
+  // results
+  constexpr DangerousCopyAssign& operator=(DangerousCopyAssign& a) {
+    ++a.val;
+    this->val = a.val;
+    return *this;
+  }
+
+  // safe copy assignment std::as_const inside ranges::iota should ensure this
+  // overload gets called
+  constexpr DangerousCopyAssign& operator=(DangerousCopyAssign const& a) {
+    this->val = a.val;
+    return *this;
+  }
+
+  constexpr bool operator==(DangerousCopyAssign const& rhs) { return this->val == rhs.val; }
+
+  // prefix
+  constexpr DangerousCopyAssign& operator++() {
+    ++(this->val);
+    return *this;
+  }
+
+  // postfix
+  constexpr DangerousCopyAssign operator++(int) {
+    auto tmp = *this;
+    ++this->val;
+    return tmp;
+  }
+};
+
+template <class Iter, class Sent, std::size_t N>
+constexpr void test_result(std::array<int, N> input, int starting_value, std::array<int, N> const expected) {
+  { // (iterator, sentinel) overload
+    auto in_begin = Iter(input.data());
+    auto in_end   = Sent(Iter(input.data() + input.size()));
+    std::same_as<std::ranges::out_value_result<Iter, int>> decltype(auto) result =
+        std::ranges::iota(std::move(in_begin), std::move(in_end), starting_value);
+    assert(result.out == in_end);
+    assert(result.value == starting_value + static_cast<int>(N));
+    assert(std::ranges::equal(input, expected));
+  }
+
+  { // (range) overload
+    // inthe range overload adds the additional constraint that it must be an outputrange
+    // so skip this for the input iterators we test
+    auto in_begin = Iter(input.data());
+    auto in_end   = Sent(Iter(input.data() + input.size()));
+    auto range    = std::ranges::subrange(std::move(in_begin), std::move(in_end));
+
+    std::same_as<std::ranges::out_value_result<Iter, int>> decltype(auto) result =
+        std::ranges::iota(range, starting_value);
+    assert(result.out == in_end);
+    assert(result.value == starting_value + static_cast<int>(N));
+    assert(std::ranges::equal(input, expected));
+  }
+}
+
+template <class Iter, class Sent = sentinel_wrapper<Iter>>
+constexpr void test_results() {
+  // Empty
+  test_result<Iter, Sent, 0>({}, 0, {});
+  // 1-element sequence
+  test_result<Iter, Sent, 1>({1}, 0, {0});
+  // Longer sequence
+  test_result<Iter, Sent, 5>({1, 2, 3, 4, 5}, 0, {0, 1, 2, 3, 4});
+}
+
+constexpr void test_DangerousCopyAssign() {
+  using A                   = DangerousCopyAssign;
+  using Iter                = contiguous_iterator<A*>;
+  std::array<A, 3> aa       = {A{1}, A{2}, A{3}};
+  std::array<A, 3> expected = {A{0}, A{1}, A{2}};
```

Comment:
I'm missing something. I thought the non-const assignment should _not_ be selected, but `expected` values being different from the original implies that it is?

Reply:
This is a bit confusing, it took me a couple of passes to remember what I was doing. If the non-const assignment is called, the values will from element `i` to `i+1` will change _by 2_.

To hopefully reduce future confusion, I'd added some comments describing this in `DangerousCopyAssign` and `test_dangerous_copy_assign`.

Reply:
Would it work to simply `= delete` the undesired assignment operator, so that if it's selected, there's a compiler error? If that doesn't work, then perhaps rather than doing a relatively small tweak in the undesired operator, maybe do something very noticeable, like assign some big constant (`999`, `1000`, `42`, etc.)?

Reply:
@var-const, the `delete` strategy works  (I temporarily removed the `std::as_const(__value)` in the implementation of `ranges::iota` and there was a compiler error. Since this is much cleaner, I'll go with this strategy, thanks for the suggestion!